### PR TITLE
Utils for message proxy

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

File: include/utils/image.hpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#include <string>
+
+#include <aos/common/tools/error.hpp>
+
+namespace aos::common::utils {
+
+using Digest = std::string;
+
+/**
+ * Unpacks a image archive.
```
Comment (mlohvynenko) [inline include/utils/image.hpp:20]:
```suggestion
 * Unpacks an image archive.
```

File: include/utils/image.hpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * @param digest digest string.
+ */
+RetWithError<std::string> ValidateDigest(const Digest& digest);
+
+/**
+ * Hashes the directory.
+ *
+ * @param dir directory path.
+ * @return std::string .
```
Comment (mlohvynenko) [inline include/utils/image.hpp:47]:
```suggestion
 * @return std::string.
```

File: src/utils/json.cpp [hunk 3]
```diff
@@ -88,6 +90,22 @@ aos::RetWithError<Poco::Dynamic::Var> ParseJson(std::istream& in) noexcept
...
+{
+    std::ofstream file(path);
+    if (!file.is_open()) {
+        return ErrorEnum::eFailed;
+    }
+
+    try {
+        Poco::JSON::Stringifier::stringify(json, file);
+    } catch (const std::exception& e) {
+        return ErrorEnum::eRuntime;
```
Comment (mlohvynenko) [inline src/utils/json.cpp:103]: It's worth adding log with error that will describe the reason of the failure

File: include/utils/filesystem.hpp [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+#include <string>
+
+#include <aos/common/tools/error.hpp>
+
+namespace aos::common::utils {
+/*
+ * Creates a temporary directory.
+ *
+ * @param dir directory path.
+ * @param pattern directory name pattern.
```
Comment (mykola-kobets-epam) [inline include/utils/filesystem.hpp:20]: please add more comments how the pattern look like

File: include/utils/filesystem.hpp [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ * @return aos::RetWithError<std::string> .
+ */
+RetWithError<std::string> MkdirTemp(const std::string& dir = "", const std::string& pattern = "");
```
Comment (mykola-kobets-epam) [inline include/utils/filesystem.hpp:23]: I would suggest renaming to MkTmpDir

**Review by mykola-kobets-epam — APPROVED** [review]

File: include/utils/image.hpp (hunk 1 above)
Comment (MykolaSolyanko) [reply include/utils/image.hpp]: Done

File: include/utils/image.hpp (hunk 2 above)
Comment (MykolaSolyanko) [reply include/utils/image.hpp]: Done

File: src/utils/json.cpp (hunk 3 above)
Comment (MykolaSolyanko) [reply src/utils/json.cpp]: Done

File: include/utils/filesystem.hpp (hunk 4 above)
Comment (MykolaSolyanko) [reply include/utils/filesystem.hpp]: Done

File: include/utils/filesystem.hpp (hunk 5 above)
Comment (MykolaSolyanko) [reply include/utils/filesystem.hpp]: Done

File: include/utils/image.hpp [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+ */
+RetWithError<std::string> UnpackTarImage(const std::string& archivePath, const std::string& destination);
+
+/**
+ * Parses the digest string.
+ *
+ * @param digest digest string.
+ * @return std::pair<std::string, std::string> .
+ */
+std::pair<std::string, std::string> ParseDigest(const std::string& digest);
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply include/utils/image.hpp:34]: decide to left as is

File: src/utils/image.cpp [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+#include <Poco/Process.h>
+#include <Poco/SHA2Engine.h>
+#include <Poco/StreamCopier.h>
+
+#include "utils/image.hpp"
+
+namespace fs = std::filesystem;
+
+namespace {
+const std::unordered_map<std::string, std::regex> anchoredEncodedRegexps = {{"sha256", std::regex(R"(^[a-f0-9]{64}$)")},
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/utils/image.cpp]: Done

File: include/utils/channel.hpp [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    RetWithError<T> Receive()
+    {
+        std::unique_lock<std::mutex> lock(mMutex);
+
+        mCond.wait(lock, [this] { return !mQueue.empty() || mExit; });
+        if (mExit) {
+            return {{}, ErrorEnum::eWrongState};
+        }
+
+        auto value = mQueue.front();
```
Comment (mlohvynenko) [inline include/utils/channel.hpp:72]:
```suggestion
        auto value = std::move(mQueue.front());
```

**Review by mlohvynenko — APPROVED** [review]

File: include/utils/channel.hpp (hunk 8 above)
Comment (MykolaSolyanko) [reply include/utils/channel.hpp]: Done

**Review by mlohvynenko — APPROVED** [review]

File: src/utils/pkcs11helper.cpp [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+#include "utils/pkcs11helper.hpp"
+
+namespace aos::common::utils {
+
+/***********************************************************************************************************************
+ * Statics
+ **********************************************************************************************************************/
+
+// The PKCS #11 URI Scheme: https://www.rfc-editor.org/rfc/rfc7512.html
+static RetWithError<std::string> CreateRFC7512URL(
```
Comment (al1img) [inline src/utils/pkcs11helper.cpp:19]: This part is updated, please rebase.

File: src/utils/filesystem.cpp [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+namespace fs = std::filesystem;
+
+namespace aos::common::utils {
+
+/***********************************************************************************************************************
+ * Static
+ **********************************************************************************************************************/
+
+static std::string JoinPath(const std::string& dir, const std::string& name)
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/utils/filesystem.cpp]: Done

File: src/utils/filesystem.cpp [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+RetWithError<std::string> MkTmpDir(const std::string& dir, const std::string& pattern)
+{
+    std::string directory            = dir.empty() ? fs::temp_directory_path().string() : dir;
+    auto [prefix, suffix]            = SplitPattern(pattern);
+    Poco::UUIDGenerator& uuidGen     = Poco::UUIDGenerator::defaultGenerator();
+    constexpr size_t     maxAttempts = 10000;
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/utils/filesystem.cpp]: removed

File: src/utils/filesystem.cpp [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    prefix = pattern;
+
+    return {prefix, suffix};
+}
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+RetWithError<std::string> MkTmpDir(const std::string& dir, const std::string& pattern)
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/utils/filesystem.cpp:27]: Done, used this function

File: src/utils/pkcs11helper.cpp (hunk 9 above)
Comment (MykolaSolyanko) [reply src/utils/pkcs11helper.cpp]: Done

**Review by al1img — APPROVED** [review]

**Pull request merged**